All peers report identical blockchains that contain duplicated transactions

<!-- For general purpose questions, use Stack Overflow http://stackoverflow.com/questions/tagged/hyperledger -->
## Description

<!-- Describe your issue or user story in detail -->

~~This is a very rare failure mode that I have only seen 3 or 4 times now. This
is the reason that I added the "duplicate check" to the "checkAgreement"
script.~~

This error appears to be 100% repeatable on my system when run as described under **How to Reproduce**, always showing 11 to 15 duplicates.

We see from the busywork/counters/driver script that all peers appear to be
working perfectly : The counts are correct and the blockchains are
identical. However, it turns out in this case that 11 transactions have been
duplicated at the end of every blockchain. Note that the duplicates come from
all over the place. (I should say that based on the responses to the
chain/blocks API it would _appear_ that the transactions were duplicated, but
I have no other knowledge of the actual internal state of the system.)

I've attached the contents of the BUSYWORK_HOME directory. Unfortunately
this run did not have DEGUG-level logs.

[busywork.tar.gz](https://github.com/hyperledger/fabric/files/324446/busywork.tar.gz)

```
network: ../bin/userModeNetwork -batch 4
network: Creating a pristine environment
network:     Killing all peer and membersrvc processes
network:     Killing all running and stopped Docker containers
network:     Removing all chaincode containers
network:     Cleaning out BUSYWORK_HOME
network:     Cleaning out /var/hyperledger
network: Running with PBFT Batch consensus
network: Network Port and PID Map:
network: ------------------------------------------------------------
network:  PeerID |  gRPC  |  REST | Events |  CLI  | Profile |  PID 
network: ------------------------------------------------------------
network:     vp0 |  38333 | 33346 |  45344 | 44592 |  45596  | 38178
network:     vp1 |  42073 | 33389 |  41646 | 42326 |  41317  | 38182
network:     vp2 |  41149 | 38904 |  46806 | 37063 |  35758  | 38190
network:     vp3 |  45836 | 46428 |  35879 | 40957 |  42482  | 38198
network: ------------------------------------------------------------
network: Waiting for peers to start
network: Network is up
driver: ./driver -clients 1 -transactions 16384 -peerBurst 10 -netBurst 10 -interlock -keepLog -checkAgreement -dupCheck
driver: List of targeted peers with REST API ports
driver:     0.0.0.0:33346
driver:     0.0.0.0:33389
driver:     0.0.0.0:38904
driver:     0.0.0.0:46428
driver: 1 chaincode(s) will be deployed
driver: 1 array(s) will be defined per chaincode
driver: 1 clients will be activated
driver: 16384 transaction(s) will be issued to each array
driver: 16384 total transaction(s) will be issued
driver: busywork::Logger : Logging to /home/bcbrock/.busywork/fabricLog
driver: The log file will persist after exit
driver: Security is disabled
driver: Deploying chaincode(s)
driver: Deployed cc0 as 2ee2fc5851a4c162bc728482c3b1e2d1bb965ab3eca6eb0bd9bb65e8df23677de64f822c6a936310105920a9f34718a9a1b272f3aee0db3c28a27cd50f87b7ef
driver: Waiting up to 5m for chaincode deployments to be registered
driver: Waiting up to 5m for all chaincode containers to be running
driver: Waiting 10s to give chaincode(s) time to initialize
driver: Waiting up to 60s for array creation to be registered
driver: Spawning clients:
driver:     Client 0 is subprocess 44863 with seed 280399412
driver: Waiting (indefinitely) for subprocesses to complete
driver: Transaction rate : 648.35 per second (16384 / 25.270196)
driver: Waiting up to 10s for block-height consensus
driver: Checking consistency of the final state
driver: Checking for agreement between the peers
checkAgreement: Blockchain logs will be written into /home/bcbrock/.busywork
checkAgreement: Logging 0.0.0.0:33346 into /home/bcbrock/.busywork/0.0.0.0:33346.chain
checkAgreement: Logging 0.0.0.0:33389 into /home/bcbrock/.busywork/0.0.0.0:33389.chain
checkAgreement: Logging 0.0.0.0:38904 into /home/bcbrock/.busywork/0.0.0.0:38904.chain
checkAgreement: Logging 0.0.0.0:46428 into /home/bcbrock/.busywork/0.0.0.0:46428.chain
checkAgreement: Checking agreement
checkAgreement: Checking for duplicates in file /home/bcbrock/.busywork/0.0.0.0:46428.chain
checkAgreement:   8201 Blocks; 1 Deploys; 16396 Invokes
checkAgreement:ERROR:   Duplicate Report : 11 Duplicated Transactions
checkAgreement:ERROR:     <TX ID>                              <Line Numbers>
checkAgreement:ERROR:     4d2aa7d6-82ea-42f5-80af-71631abd3693 847 24582
checkAgreement:ERROR:     ad68e796-8933-418b-b0ad-91249bdb0c74 1339 24583
checkAgreement:ERROR:     cb8097e1-48f8-4bc6-8fba-60d58778042a 4012 24585
checkAgreement:ERROR:     f8496d09-c4e5-4f76-a1ac-b58b2c989ddc 4042 24586
checkAgreement:ERROR:     b392f217-f132-452f-96d3-3fc6ce620d31 4435 24588
checkAgreement:ERROR:     6ca279dd-2e1f-4943-93d2-dde763e89ba9 8305 24589
checkAgreement:ERROR:     e7f2b363-0c10-4eb6-8a8a-29a207e2b518 9781 24591
checkAgreement:ERROR:     d6d7aefa-3850-4568-94dd-06eb6870a0e4 10594 24592
checkAgreement:ERROR:     6a9c3884-20bc-4f79-acbb-e523db9faf86 16675 24594
checkAgreement:ERROR:     df94dd1a-a140-4af8-a102-a67fb36eb4f4 1987 24596
checkAgreement:ERROR:     e3659e4b-6d30-4bd7-a537-7c4e12d2ceae 12076 24597
checkAgreement:ERROR: Aborting due to errors, disgreement or duplicates; All log files have been kept
driver:ERROR: Peer agreement check failed
driver:ERROR: Aborting due to errors, mismatches or disagreements above
private.mk:3: recipe for target 's2b' failed
```
## Describe How to Reproduce

<!-- If an issue, provide sufficient context and steps to reproduce the issue -->

This issue was observed on an X86 server running

```
commit 6d873142362f2e6314ca138defa792dcc52b4f10
Merge: 77f7077 6ea2c1e
Author: Binh Q Nguyen <binhn@us.ibm.com>



Date:   Mon Jun 20 11:28:41 2016 -0400

    Merge pull request #1912 from mrshah-at-ibm/chaincode_tls

    Chaincode tls
```

The busywork Make target is below; This failure occurred on the first test (single client). Note that I am using busywork test code that has not been merged yet.

```
.PHONY: s2b 
s2b:
    for clients in 1 2 4 8 16 32; do \
    CORE_PBFT_GENERAL_TIMEOUT_REQUEST=5s \
    CORE_PBFT_GENERAL_TIMEOUT_VIEWCHANGE=5s \
    CORE_PBFT_GENERAL_VIEWCHANGEPERIOD=0 \
    CORE_PBFT_GENERAL_NULLREQUEST=1s \
    $(NETWORK) -batch 4; sleep 10; \
    set -e; ./driver \
        -clients $$clients \
        -transactions $$((16384 / $$clients)) \
        -peerBurst 10 -netBurst 10 \
        -interlock \
        -keepLog \
        -checkAgreement -dupCheck; \
    done
```

--- a/consensus/obcpbft/obc-batch.go
+++ b/consensus/obcpbft/obc-batch.go
@@ -50,6 +50,8 @@ type obcBatch struct {
 
 	reqStore *requestStore // Holds the outstanding and pending requests
 
+	deduplicator *deduplicator
+
 	persistForward
 }
 
@@ -104,6 +106,8 @@ func newObcBatch(id uint64, config *viper.Viper, stack consensus.Stack) *obcBatc
 	op.batchTimer = etf.CreateTimer()
 
 	op.reqStore = newRequestStore()
+
+	op.deduplicator = newDeduplicator()
 
 	op.idleChan = make(chan struct{})
 	close(op.idleChan) // TODO remove eventually
@@ -213,6 +217,8 @@ func (op *obcBatch) execute(seqNo uint64, raw []byte) {
 			logger.Debugf("Batch replica %d missing transaction %s outstanding=%v, pending=%v", op.pbft.id, tx.Uuid, outstanding, pending)
 		}
 		txs = append(txs, tx)
+
+		op.deduplicator.Execute(req)
 	}
 
 	meta, _ := proto.Marshal(&Metadata{seqNo})
@@ -310,6 +316,11 @@ func (op *obcBatch) processMessage(ocMsg *pb.Message, senderHandle *pb.PeerID) e
 	}
 
 	if req := batchMsg.GetRequest(); req != nil {
+		if !op.deduplicator.IsNew(req) {
+			logger.Warningf("Replica %d ignoring request as it is too old", op.pbft.id)
+			return nil
+		}
+
 		if (op.pbft.primary(op.pbft.view) == op.pbft.id) && op.pbft.activeView {
 			return op.leaderProcReq(req)
 		}

